Refactoring

peterpaints opened the pull request.

peterpaints committed:
feat: Create Users and Sessions controllers

peterpaints committed:
feat: Create Login/Register pages

- Add phone_number and slack_handle columns to db
- Add rubocop rules, excluding boilerplate files
- Add login/register routes

[Delivers #159712136]

peterpaints committed:
Merge pull request #1 from peterpaints/ft-authentication-159712136

feat: Create Login/Register pages

peterpaints committed:
feat: Create rides controller

peterpaints committed:
feat: Create template for rides view

- Create template(s) for user profile settings
- Add flash messages bootstrap alert
- Restrict logged in user to own profile page

peterpaints committed:
Merge pull request #2 from peterpaints/ft-user-profile-159688105

Ft user profile 159688105

peterpaints committed:
feat: Create Ride and Vehicle models

peterpaints committed:
feat: Add has_many and belongs_to relationships

peterpaints committed:
feat: Add association migrations

peterpaints committed:
feat: Create join table for users and rides

peterpaints committed:
Merge pull request #3 from peterpaints/ft-vehicle-ride-models-159684828

Ft vehicle ride models 159684828

peterpaints committed:
chore: Update README.md and include license

peterpaints committed:
feat: Clean up authentication

- Animation for flash messages
- Protect certain routes with require_login

peterpaints committed:
feat: Simplify navigation

- Add tabs to settings page
- Label pages more clearly

peterpaints committed:
feat: Add edit and delete functionality on frontend, for rides and vehicles

peterpaints committed:
feat: Implement CRUD functionality for Vehicles

peterpaints committed:
feat: Add CRUD functionality for Rides

peterpaints committed:
feat: Finish Booking functionality

peterpaints committed:
feat: Allow users to edit notification preferences

peterpaints committed:
Merge pull request #4 from peterpaints/ft-controllers-functionality-159684827

Ft controllers functionality 159684827

peterpaints committed:
feat: Add tests for the models

peterpaints committed:
chore: Set up circleci

peterpaints committed:
Add tests for UsersController

peterpaints committed:
fix: Modify simplecov config to exclude config and spec folders

peterpaints committed:
chore: Add SessionsController tests

peterpaints committed:
fix: Update UsersController tests to raise cov

peterpaints committed:
chore: Add VehiclesController tests

peterpaints committed:
chore: Add RidesController tests

peterpaints committed:
chore: Deploy to heroku

peterpaints committed:
fix: Remove turbolinks to allow jQuery to load on heroku

peterpaints committed:
fix: Refactor navbar links to work on FireFox

peterpaints committed:
fix: Prevent notifications badge moving when flash message shows

- Add paragraph when user vehicle/rides are empty.

peterpaints committed:
chore: Separate ride callback methods

peterpaints committed:
fix: Remove repeated user factory calls

peterpaints committed:
fix: Center notification divs

peterpaints committed:
fix: On initial load after register, show tab content in correct tabs

- The tab would default to `your vehicles` and show every other tab content

peterpaints committed:
Merge pull request #5 from peterpaints/ft-tests-auth-controllers-models-159684840

Ft tests auth controllers models 159684840

peterpaints committed:
feat: Add notification functionality

- Users receive notifications when core ride actions occur. E.g. booking, deletion, etc.
- Flash message is now better placed. On the side. No longer moves page down.
- Users can now filter available rides
- Remove password validation in login form - unnecessary
- Remove filter buttons from user settings page - unnecessary

peterpaints committed:
chore: Add more validation to Rides

- Can't have capacity > vehicle capacity
- Can't create ride in the past

peterpaints committed:
feat: Add search functionality

peterpaints committed:
fix: Undo vehicle capacity faker as it fails VehicleOverload check

peterpaints committed:
feat: Add tooltip to show riders

- Extract filter functionality into own module.

peterpaints committed:
fix: Prevent user from lowering capacity of existing booked rides below number of riders

peterpaints committed:
Merge pull request #6 from peterpaints/ft-notifications-159684843

Ft notifications 159684843

peterpaints committed:
fix: Add includes(:users) to rides query for optimization

peterpaints committed:
chore: Add notifications tests

peterpaints committed:
chore: Add search icon to search bar

peterpaints committed:
fix: Make all input fields title case

hmasila commented on Rakefile:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
you can exclude rakefile as well

peterpaints replied:
Done

hmasila commented on app/controllers/concerns/filter_search.rb:
```diff
@@ -0,0 +1,18 @@
+  def filter_rides(params)
+    @rides = @rides.reverse if params[:filter] == 'asc'
+    if params[:filter] == 'mine'
+      @rides = @rides.joins(:users).where(users: { id: @user.id })
```
this seems more of logic that should be in the models. Maybe as a scope in the ride model

hmasila replied:
You don't need the`joins` query if there is a relationship defined between the two models. In this case, `@rides.find_by(user_id: @user.id)`

peterpaints replied:
I tried this but it failed. Any ideas. For now I've left it as is. But it's in a scope now

hmasila commented on app/controllers/notifications_controller.rb:
```diff
@@ -0,0 +1,37 @@
+
+  private
+
+  def find_current_user
```
What is the difference between this method and `current_user` method

hmasila commented on app/controllers/notifications_controller.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def mark_as_read
+    @user.notifications.where(read: false).each do |notification|
```
This is business logic that should be in the models. Maybe in the user model. The method here should just call that method but not interact directly with the database

hmasila commented on app/controllers/rides_controller.rb:
```diff
@@ -0,0 +1,124 @@
+  rescue_from Exceptions::RiderLacksSeating, with: :rider_lacks_seating_error
+
+  def index
+    @rides = Ride.includes(:users).where('departure_time > ?', Time.now)
```
Make use of scopes. This is logic that should be in the models

hmasila commented on app/controllers/rides_controller.rb:
```diff
@@ -0,0 +1,124 @@
+  before_action :find_ride, only: %i[book cancel update destroy]
+  before_action :find_current_user, only: %i[index book cancel]
+
+  rescue_from Exceptions::OwnerBooking, with: :ride_booking_error
```
You can create a module or class to take care of the exceptions and just include it in the application controller so that it is accessible to all the other controllers.

Check out https://github.com/hmasila/buclist/blob/develop/app/controllers/concerns/exception_handler.rb

hmasila commented on app/controllers/rides_controller.rb:
```diff
@@ -0,0 +1,124 @@
+  def create
+    @ride = Ride.new(ride_params)
+    if @ride.save
+      ride_save_success
```
Nice abstracting this

hmasila commented on app/controllers/rides_controller.rb:
```diff
@@ -0,0 +1,124 @@
+  end
+
+  def find_current_user
+    @user = User.find(current_user)
```
What are the contents of `current_user` and how different are they from the `find_current_user` method

peterpaints replied:
I've changed current_user to return an instance instead of just the id

hmasila commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,31 @@
+
+  def create
+    @user = User.find_by_email(params[:email])
+    if @user&.authenticate(params[:password])
```
+1 on use of the navigation operator

hmasila commented on app/models/concerns/exceptions.rb:
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+module Exceptions
```
Nice. Now just include the rescues here. REF https://github.com/hmasila/buclist/blob/develop/app/controllers/concerns/exception_handler.rb

hmasila commented on app/models/ride.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def cannot_book_full_ride(_user)
+    raise RideFull if users.size == capacity
```
Can there be a time when `users.size > capacity`

peterpaints replied:
Yeah, but it stops you from booking a ride if that happens as well

hmasila commented on config/initializers/assets.rb:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
you can exclude `config` from rubocop

peterpaints replied:
Done

hmasila commented on spec/controllers/notifications_controller_spec.rb:
```diff
@@ -0,0 +1,30 @@
+  describe 'POST #create' do
+    context 'create request is valid' do
+      before do
+        stub_current_user(user)
```
+1 on stubbing the user

hmasila replied:
lol, reading that loud sounds wrong haha

hmasila commented on spec/controllers/notifications_controller_spec.rb:
```diff
@@ -0,0 +1,30 @@
+      get :mark_as_read, xhr: true
+    end
+
+    it 'notification to be marked as read' do
```
Try and make the specs sound more English like. For example here, you could have `it 'marks the notification as read'`

hmasila commented on spec/controllers/vehicles_controller_spec.rb:
```diff
@@ -0,0 +1,118 @@
+require 'rails_helper'
+
+RSpec.describe VehiclesController, type: :controller do
+  let(:user) { create(:user) }
```
This can be written as `let!(:user) { create(:user) }` if you don't anticipate it to change in the course of testing

hmasila commented on spec/controllers/vehicles_controller_spec.rb:
```diff
@@ -0,0 +1,118 @@
+  end
+
+  describe 'PUT #update' do
+    let(:vehicle) { create :vehicle }
```
Use `let!` if the object is not going to change within the scope

hmasila commented on spec/controllers/vehicles_controller_spec.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+
+      it 'shows flash error' do
+        expect(assigns(:vehicle).errors[:capacity]).to include 'is invalid'
```
Nice work on the use of assigns. As for the message, you can also play around with matches and use regex -- not necesarily but it's a good to have knowledge

hmasila commented on spec/factories/notification.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
```
Goos use of FactoryBot 🎉

hmasila commented on spec/factories/notification.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :notification do
+    user
+    message { Faker::SiliconValley.quote }
```
Great work using Fakers 🎉

hmasila commented on spec/models/ride_spec.rb:
```diff
@@ -0,0 +1,28 @@
+  context 'validations' do
+    taken = 'That ride already exists.'
+
+    it { is_expected.to validate_presence_of(:origin) }
```
Nice work using Shoulda matchers 🎉

hmasila commented on spec/models/ride_spec.rb:
```diff
@@ -0,0 +1,28 @@
+    it { is_expected.to validate_presence_of(:capacity) }
+    it { is_expected.to validate_presence_of(:vehicle_id) }
+
+    it do
```
What just happened here? Insert :flashed: emoji

hmasila commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,29 @@
+    it { is_expected.to allow_value('user@example.com').for(:email) }
+    it { is_expected.not_to allow_value('user.com').for(:email) }
+
+    it { is_expected.to have_secure_password }
```
Awesome. Simply awesome job testing for edge cases 🎉

hmasila commented on spec/support/helpers.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module Helpers
+  def stub_current_user(user)
```
👍

hmasila reviewed (commented):
Great work!! I have left some comments here and there. The most significant feedback is keep your controllers as thin as possible. Move all the business and database logic to the models. Make use of scopes in the models

peterpaints committed:
fix: Refactor with mentor-advised changes

- Extract error handling in ride controller to concern
- Extract error raising in ride model to concern
- Change the current_user method to return an instance of user instead of id
- Add more scopes to the Ride Model

peterpaints committed:
fix: Users now see ride deletion notifications
